compiler/noirc_evaluator/src/ssa/ir/dfg.rs

```diff
@@ -21,6 +22,75 @@ use serde::{Deserialize, Serialize};
 use serde_with::serde_as;
 use serde_with::DisplayFromStr;
 
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub(crate) struct CallStackId(u32);
+
+impl CallStackId {
+    pub(crate) fn root() -> Self {
+        Self::new(0)
+    }
+
+    fn new(id: usize) -> Self {
+        Self(id as u32)
+    }
+
+    pub(crate) fn index(&self) -> usize {
+        self.0 as usize
+    }
+
+    pub(crate) fn is_root(&self) -> bool {
+        self.0 == 0
+    }
+
+    // Retrieves a CallStack from a CallStackId
+    fn get_call_stack(&self, locations: &[LocationNode]) -> CallStack {
+        let mut call_stack = List::new();
+        let mut current_location = *self;
+        while let Some(parent) = locations[current_location.index()].parent {
+            call_stack.push_back(locations[current_location.index()].value);
+            current_location = parent;
+        }
+        call_stack
+    }
+
+    // Adds a location to the call stack
+    fn add_location(&self, location: Location, locations: &mut Vec<LocationNode>) -> CallStackId {
+        if let Some(result) = locations[self.index()]
+            .children
+            .iter()
+            .find(|child| locations[child.index()].value == location)
```

Comment:
There are quite a few invariants assumed about the `locations` vector, e.g. that `self` is in it, and that the `children` of a `LocationNode` are all in it as well; it would be worth creating a _newtype_ to wrap the `Vec` and make sure these invariants hold, and you're not dealing with just any odd `vec!` someone made.